### ISSUE TITLE
action recognition testing

### ISSUE THREAD

**Issue by Drogba78** [dmlc/gluon-cv#1304]
> For testing (reporting these numbers), we use 250 views for 2D networks (25 frames and 10-crop).

So, when I do testing for 2D networks, I set `num_segments=25` and `num_crops=10`, is it correct?
@bryanyzhu

**Comment by yezqNLP** [comment]
During training, I set `num_segments=3` and `num_crops=1`. I found the evaluation accuracy and test accuracy are almost same. However, with more crops and more segments, I think test accuracy should be higher.

**Comment by bryanyzhu** [comment]
Usually yes, with more crops and more segments, the test accuracy is higher. But it also depends on specific network architecture. Some networks benefit more from this multi-view testing (e.g., about 10 percent improvement), some benefit less (e.g., about 2 percent improvement). As long as you can reproduce the final test number as our reported numbers in the model zoo, I think it is fine.

**Comment by yezqNLP** [comment]
> Usually yes, with more crops and more segments, the test accuracy is higher. But it also depends on specific network architecture. Some networks benefit more from this multi-view testing (e.g., about 10 percent improvement), some benefit less (e.g., about 2 percent improvement). As long as you can reproduce the final test number as our reported numbers in the model zoo, I think it is fine.

Well, for resnet50_v1b_hmdb51 model with `num_segments=3`, during training I got eval acc about ~0.45, and during testing I got test acc ~0.46. However, the reported acc is 0.55 in the tutorial. It might relate to the batch_size that I use, I use 2 gpus (each with 11G), during training, batch_size is set to 16; and for testing, as the `num_of_crops` and `num_of_segments` is large, batch_size is set to 2. 
To sum up, the differences I found out are `num_of_gpu` and `batch_size` as limited memory, which might cause the performance drop. Can u give me any advice? Also, can you share your test_script.sh? I might have some wrong configuration in test_script.sh.

**Comment by bryanyzhu** [comment]
For training, I think 2 GPUs each with 11G should be good for ResNet50 model. For testing, due to large `num_crops` and `num_segments`, the batch size should be 1 per GPU. 

Here is my script for testing,

```
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 python ./scripts/action-recognition/test_recognizer.py \
    --dataset hmdb51 \
    --data-dir /home/ubuntu/.mxnet/datasets/hmdb51/rawframes \
    --val-list /home/ubuntu/.mxnet/datasets/hmdb51/testTrainMulti_7030_splits/hmdb51_val_split_1_rawframes.txt \
    --model resnet50_v1b_hmdb51 \
    --use-pretrained \
    --num-classes 51 \
    --mode hybrid \
    --dtype float32 \
    --prefetch-ratio 1.0 \
    --batch-size 1 \
    --num-gpus 8 \
    --num-data-workers 32 \
    --new-height 256 \
    --new-width 340 \
    --new-length 1 \
    --new-step 1 \
    --ten-crop \
    --input-size 224 \
    --num-segments 25 \
    --log-interval 10 \
```

**Comment by yezqNLP** [comment]
Update, after check your script, I think my script is fine. 
When I use `--use-pretrained`, I can get ~0.55 testing acc, which is reported in the tutorial. 
However, if I load my saved parameters, still the evaluation acc and testing acc is almost the same. And I tried several models, like resnet50, resent18, both in hmdb51 and ucf101. For example, in hmdb51 dataset, I re-run the resnet50 model, with `num_segments=3, batch_size=32`, I got evaluation acc ~0.517 and testing acc ~0.513. 
I wonder If u get time to check your evaluation acc for `resnet50_v1b_hmdb51` with `num_segments=3, batch_size=32` in the tensorboard, to find out evaluation acc is close to 0.55 or not.

**Comment by bryanyzhu** [comment]
You can find my training log at here, https://raw.githubusercontent.com/dmlc/web-data/master/gluoncv/logs/action_recognition/hmdb51/resnet50_v1b_hmdb51_tsn.log

I think during evaluation, our acc is also close to 0.55.

**Comment by yezqNLP** [comment]
> You can find my training log at here, https://raw.githubusercontent.com/dmlc/web-data/master/gluoncv/logs/action_recognition/hmdb51/resnet50_v1b_hmdb51_tsn.log
> 
> I think during evaluation, our acc is also close to 0.55.

My acc only reaches ~0.51, I think the performance drop mainly comes from `num_of_gpu` used. As I found when I compare `num_of_gpu=1` and `num_of_gpu=2`, 2 gpus have better performance than 1 gpu.

**Comment by yezqNLP** [comment]
So, I try to use `gradient accumulation` to have larger batch_size. However, I found that this code in train_recognizer.py may some problem. 

```
if opt.accumulate > 1 and (i + 1) % opt.accumulate == 0:
    if opt.kvstore is not None:
        trainer.step(batch_size * kv.num_workers * opt.accumulate)
    else:
        trainer.step(batch_size * opt.accumulate)
        net.collect_params().zero_grad()
else:
    if opt.kvstore is not None:
        trainer.step(batch_size * kv.num_workers)
    else:
        trainer.step(batch_size)
```

I think it should be like this 
```
if opt.accumulate > 1:
    if (i + 1) % opt.accumulate == 0:
        if opt.kvstore is not None:
            trainer.step(batch_size * kv.num_workers * opt.accumulate)
        else:
            trainer.step(batch_size * opt.accumulate)
            net.collect_params().zero_grad()
else:
    if opt.kvstore is not None:
        trainer.step(batch_size * kv.num_workers)
    else:
        trainer.step(batch_size)
```

**Comment by bryanyzhu** [comment]
Hi, I didn't see any difference between the code snippets you provided above. I think they are equivalent, please correct me if I am wrong.

**Comment by bryanyzhu** [comment]
> > You can find my training log at here, https://raw.githubusercontent.com/dmlc/web-data/master/gluoncv/logs/action_recognition/hmdb51/resnet50_v1b_hmdb51_tsn.log
> > I think during evaluation, our acc is also close to 0.55.
> 
> My acc only reaches ~0.51, I think the performance drop mainly comes from `num_of_gpu` used. As I found when I compare `num_of_gpu=1` and `num_of_gpu=2`, 2 gpus have better performance than 1 gpu.

This is possible. If number of GPUs are small, the batch size maybe too small. This issue may be fixable by partial BN. However, I didn't implement partial BN for ResNet models. You can try partial BN following the code here (https://github.com/dmlc/gluon-cv/blob/master/scripts/action-recognition/train_recognizer.py#L413-L416).

**Comment by yezqNLP** [comment]
> Hi, I didn't see any difference between the code snippets you provided above. I think they are equivalent, please correct me if I am wrong.


Well, the difference happens when accumulate>1. 
To make it simple, I remove the `kvstore` related code. 
Suppose accumulate=4, and when i=4,5,6, we do not it execute line 1 `trainer.step(batch_size * opt.accumulate)` or line 4 `trainer.step(batch_size)`. however, line 4 `trainer.step(batch_size)` will be executed. correct me if I am wrong.

```
0 if opt.accumulate > 1 and (i + 1) % opt.accumulate == 0:
1    trainer.step(batch_size * opt.accumulate)
2    net.collect_params().zero_grad()
3 else:
4    trainer.step(batch_size)
```

**Comment by bryanyzhu** [comment]
> > Hi, I didn't see any difference between the code snippets you provided above. I think they are equivalent, please correct me if I am wrong.
> 
> Well, the difference happens when accumulate>1.
> To make it simple, I remove the `kvstore` related code.
> Suppose accumulate=4, and when i=4,5,6, we do not it execute line 1 `trainer.step(batch_size * opt.accumulate)` or line 4 `trainer.step(batch_size)`. however, line 4 `trainer.step(batch_size)` will be executed. correct me if I am wrong.
> 
> ```
> 0 if opt.accumulate > 1 and (i + 1) % opt.accumulate == 0:
> 1    trainer.step(batch_size * opt.accumulate)
> 2    net.collect_params().zero_grad()
> 3 else:
> 4    trainer.step(batch_size)
> ```
Oh, that's True. I think you are right. You can use your gradient accumulation to see if you can obtain good results on HMDB51.

**Comment by yezqNLP** [comment]
One more thing, I think the following should be changed accordingly:
https://github.com/dmlc/gluon-cv/blob/3c4150a964c776e4f7da0eb30b55ab05b7554c8d/scripts/action-recognition/train_recognizer.py#L341

```
num_batches = len(train_data) // opt.accumulate
```

**Comment by bryanyzhu** [comment]
Let me know if you can achieve good results using your fix, and welcome to submit a PR to fix these issues. Thank you.

**Comment by Drogba78** [comment]
For resnet50 model with `num_segments=3`, `batch_size=32`, I got evaluation acc ~0.517
After I set `accumulate=4`, the acc reaches to ~0.54.

And I submit a PR to fix the accumulate problem.
https://github.com/dmlc/gluon-cv/pull/1313

**Comment by bryanyzhu** [comment]
Good to know you get better results using accumulation. And thank you for your contribution, I just merged your PR.

**Comment by Drogba78** [comment]
> Good to know you get better results using accumulation. And thank you for your contribution, I just merged your PR.

Thanks for your quick reply, too : ) .
